### ISSUE TITLE
Backburner — GELU epilogue fusion for rec (26% pool, custom-kernel work)

### ISSUE THREAD

**Issue by vgf89** [vgf89/InstantJPDict#27]
## Parent

Follow-on from #25 (perf research). Not scheduled — parked here so the analysis isn't lost.

## Finding

Per-layer profile (Pixel 7a, rec w480, 1-thread): GELU = 26% of inference (~18ms of 68ms), 13 layers at ~1.5ms each computing erf-flavored tanh. Both cheap fixes are closed:

- ncnn has **no fused-GELU activation type** (`activation_ss` ends at HardSwish; conv `9=` cannot express GELU).
- The GELU layer **already runs the tanh approximation**, not erf — no fast-math juice left.
- The 13 producers are int8 1x1s whose ARM kernels (`im2col_gemm_int8` / `packed_int8`) have **no activation epilogue at all**.

## The actual work

Add GELU epilogues to the ARM int8 conv paths per packing variant (pack1/4/8) in the vendored tree (`/home/holopengin/repos/ncnn`, no PR per owner policy), add a fused-activation type, then graph surgery (set fused type on the 13 producers, delete the GELU layers, renumber blobs). Gate: host CER parity + device min-of-3 A/B.

## Why backburner

Days of kernel work with pack-variant regression surface for ~10–18ms. Revisit if easier pools dry up.

*penned by opencode + Muse Spark*

**Comment by vgf89** [comment]
Claiming — pools dry per maintainer, starting verification first (confirm no fused-GELU type, int8 epilogue absence, re-measure 13-layer pool) before any kernel writes. Gate stays host CER parity + device min-of-3 A/B.

*penned by opencode + Muse Spark*

**Comment by vgf89** [comment]
Verification first pass — one correction to the issue body before any kernel work:

**Claim B is backwards.** The 13 GELU lines in rec_dyn.param are bare (no 0= field), and ncnn defaults fast_gelu=0 = erf path (gelu.cpp load_param default 0; tanh only when 0=1). So the graph runs erf today, not the tanh approx — there IS cheap fast-math juice left: setting 0=1 changes GELU output by at most 4.8e-4 over the full range (numpy), comfortably inside logit-epsilon territory.

**Claims A/C confirmed with one refinement.** No fused-GELU type (activation_ss ends at HardSwish=6); int8 GEMM tiles emit raw int32 with no epilogue. But the int8 path already has an activation slot one stage later (requantize generic fuses activation_ss/ps in-loop; fp32-out 8=2 path does dequantize + activation layer). Our 13 producers are all 8=2, so headers + graph surgery may be enough for correctness — per-pack GEMM epilogues are the full-perf version, not necessarily the first version.

**Artifacts this session:**
- tools/fuse_gelu_rec.py — --mode fast (13x GELU 0=1, minimal diff) and --mode fuse (13 producers get 9=7, GELUs deleted, blobs renumbered 179/206 -> 166/193, gemm_8 extract id 204 -> 191 for ncnn_jni.cpp).
- Both outputs structurally validated (header counts match, numeric ids contiguous from 1, no undefined inputs, consumer rewiring spot-checked).
- Host x86_64 libncnn build running for rec_textcmp parity (fast variant CER/maxAbs vs baseline); device min-of-3 A/B after.

No kernel edits yet. Next: host parity numbers, then a device A/B of the fast-only variant to see how much of the ~18ms pool the tanh approx alone recovers.

*penned by opencode + Muse Spark*

**Comment by vgf89** [comment]
Host parity for the fast_gelu-only variant (baseline int8 rec_dyn vs 13x 0=1, 60 calib files, host x86_64 libncnn @0c9625b):

- w480: exact 57/60 (95.0%), CERlike 3/708 = 0.004 — gate needs >=88% / <=0.020: PASS with margin
- w256 canary: exact 58/60 (96.7%), CERlike 2/266 = 0.008 — gate >=78% / <=0.060: PASS
- per-timestep argmax agree 3574/3600 (99.28%); logit maxAbs up to ~0.99 on 2 files (tiny GELU diffs amplify through attention, CTC stays robust); softmax-out sortedMaxAbs 0.02
- all 5 text diffs are single-edit (ed=1) on short strings

So the tanh approx is parity-safe by the project's own gate. No device attached to this session, so the speed half is yours — device A/B recipe (same-session, min-of-3, benchRecNcnnDynWidths p50 at w480):

  cp app/src/main/assets/PP-OCRv6_small_ncnn/rec_dyn.param /tmp/rec_dyn_erf.param.bak
  cp /tmp/rec_dyn_fastgelu.param app/src/main/assets/PP-OCRv6_small_ncnn/rec_dyn.param
  ./gradlew :app:assembleBenchmark
  ./gradlew :app:connectedReleaseAndroidTest -Pandroid.testInstrumentationRunnerArguments.class=com.holopengin.instantjpdict.OcrBenchmarkTest
  adb logcat -s OcrBenchmark | grep -E 'p50|SUMMARY'
  cp /tmp/rec_dyn_erf.param.bak app/src/main/assets/PP-OCRv6_small_ncnn/rec_dyn.param  # restore

Run A (erf, restored file) then B (fast) 3x same session/charger state, take min p50 per variant. If B recovers a real chunk of the ~18ms pool at CER parity, ship B as a param-only change and shrink the remaining fusion work to the true memory-pass saving. If B shows ~0ms, the erf cost was never the pool and full epilogue fusion is the only lever.

*penned by opencode + Muse Spark*

**Comment by vgf89** [comment]
Device A/B verdict — Pixel 7a (lynx), same session, debug build, benchRecNcnnDynWidths w480 (synthetic, p50 of 5):

- A baseline (erf, bare GELU): 72, 70, 70 -> min 70ms
- B fast approx (13x GELU 0=1): 60, 59, 59 -> min 59ms
- Delta: -11ms (-16%) at host CER parity (0.004 w480, 0.008 w256 — see prior comment)

So the tanh approx alone recovers ~11 of the ~18ms pool with a 13-line param-only change. Recommendation: ship B (regenerate rec_dyn from the fast variant via tools/fuse_gelu_rec.py --mode fast + full parity_gate_rec.sh re-quant check), and descope the remaining custom-kernel fusion to the leftover ~7ms — likely not worth days of pack-variant work. The fused-9=7 path (tools/fuse_gelu_rec.py --mode fuse output validated at 166/193) stays parked until easier pools dry up again.

Method notes: connectedReleaseAndroidTest does not exist in this AGP setup (only connectedDebugAndroidTest); device needed adb install -d after the release-signed build; ran via manual am instrument (7 tests OK each run). Every local assembleDebug also rewrites the checked-in app/src/main/jniLibs/arm64-v8a/libnav_graph_core.so (build nondeterminism, restored after) — separate hygiene item.

Working tree left clean on wayfinder-ncnn-port (only commit: surgery script). Device restored to baseline-erf debug build.

*penned by opencode + Muse Spark*

**Comment by vgf89** [comment]
Shipped per maintainer direction:

- Commit 1b2aca0 (wayfinder-ncnn-port): rec_dyn.param 13x GELU 0=1. Parity re-run on shipping bytes: 95.0%/CER 0.004 w480.
- Benchmark APK (assembleBenchmark) installed on the Pixel 7a just now — APK verified to contain the fast asset (13/13 GELU with 0=1, 179/206 graph).
- Remaining kernel fusion parked in #30 (fused type-7 work list + validated fuse tooling + ~7ms pool).

Leaving this open for your on-phone verification before close.

*penned by opencode + Muse Spark*

**Comment by vgf89** [comment]
Maintainer verified on-phone (feels faster). Shipped: 1b2aca0 fast-GELU asset (w480 70 -> 59ms min-of-3, CER 0.004). Remainder parked in #30.

*penned by opencode + Muse Spark*
